core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java

```diff
@@ -193,14 +294,192 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeGenericValue(points);
+        
+        geoDistance.writeTo(out);
+        unit.writeTo(out);
+        order.writeTo(out);
+        out.writeOptionalString(sortMode);
+        if (nestedFilter != null) {
+            out.writeBoolean(true);
+            out.writeQuery(nestedFilter);
+        } else {
+            out.writeBoolean(false);
+        }
+        out.writeOptionalString(nestedPath);
+        out.writeBoolean(coerce);
+        out.writeBoolean(ignoreMalformed);
+    }
+
+    @Override
+    public GeoDistanceSortBuilder readFrom(StreamInput in) throws IOException {
+        String fieldName = in.readString();
+        GeoDistanceSortBuilder result = new GeoDistanceSortBuilder(fieldName);
+
+        ArrayList<GeoPoint> points = (ArrayList<GeoPoint>) in.readGenericValue(); 
+        result.points(points.toArray(new GeoPoint[points.size()]));
+        
+        result.geoDistance(GeoDistance.readGeoDistanceFrom(in));
+        result.unit(DistanceUnit.readDistanceUnit(in));
+        result.order(SortOrder.readOrderFrom(in));
+        result.sortMode(in.readOptionalString());
+        if (in.readBoolean()) {
+            result.setNestedFilter(in.readQuery());
+        }
+        result.setNestedPath(in.readOptionalString());
+        result.coerce(in.readBoolean());
+        result.ignoreMalformed(in.readBoolean());
+        return result;
+    }
+
+    @Override
+    public GeoDistanceSortBuilder fromXContent(QueryParseContext context, String elementName) throws IOException {
+        XContentParser parser = context.parser();
+        String fieldName = null;
+        List<GeoPoint> geoPoints = new ArrayList<>();
+        DistanceUnit unit = DistanceUnit.DEFAULT;
+        GeoDistance geoDistance = GeoDistance.DEFAULT;
+        boolean reverse = false;
+        MultiValueMode sortMode = MultiValueMode.fromString(DEFAULT_SORT_MODE);
+        QueryBuilder nestedFilter = null;
+        String nestedPath = null;
+
+        boolean coerce = GeoDistanceSortBuilder.DEFAULT_COERCE;
+        boolean ignoreMalformed = GeoDistanceSortBuilder.DEFAULT_IGNORE_MALFORMED;
+
+        XContentParser.Token token;
+        String currentName = parser.currentName();
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentName = parser.currentName();
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                parseGeoPoints(parser, geoPoints);
+
+                fieldName = currentName;
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                // the json in the format of -> field : { lat : 30, lon : 12 }
+                if ("nested_filter".equals(currentName) || "nestedFilter".equals(currentName)) {
+                    nestedFilter = context.parseInnerQueryBuilder();
```

Comment:
Note to remember: while this is kept as a QueryBuilder internally, I think we need to make sure to call `toFiler()` on it once on the shard (e.g. in the new build() method, doesn't seem to be there yet)